packages/terra-application/src/application-loading-overlay/ApplicationLoadingOverlayProvider.jsx

```diff
@@ -0,0 +1,86 @@
+import React, { useMemo } from 'react';
+import PropTypes from 'prop-types';
+import classNames from 'classnames/bind';
+import Scroll from 'terra-scroll';
+import LoadingOverlay from 'terra-overlay/lib/LoadingOverlay';
+import OverlayContainer from 'terra-overlay/lib/OverlayContainer';
+
+import ApplicationLoadingOverlayContext from './ApplicationLoadingOverlayContext';
+import styles from './ApplicationLoadingOverlayProvider.module.scss';
+
+const cx = classNames.bind(styles);
+
+const propTypes = {
+  /**
+   * The components to be rendered within the context of the ApplicationLoadingOverlayProvider.
+   * Components rendered here are able to interact with ApplicationLoadingOverlayProvider through
+   * the ApplicationLoadingOverlayContext.
+   */
+  children: PropTypes.node,
+  /**
+   * A function to be called with the current ref of the scrollable element rendered within the
+   * ApplicationLoadingOverlayProvider.
+   */
+  scrollRefCallback: PropTypes.func,
+};
+
+const defaultProps = {
+  children: undefined,
+  scrollRefCallback: undefined,
+};
+
+const ApplicationLoadingOverlayProvider = ({
+  children,
+  scrollRefCallback,
+}) => {
+  const [loadingIndicators, setLoadingIndicators] = React.useState({});
+
+  const contextValue = useMemo(() => ({
+    show: (key, message) => {
+      setLoadingIndicators(state => (
+        {
+          ...state,
+          [`${key}`]: message,
+        }
+      ));
+    },
+    hide: (key) => {
+      setLoadingIndicators((state) => {
+        const newLoadingIndicators = { ...state };
+        delete newLoadingIndicators[key];
+
+        return newLoadingIndicators;
+      });
+    },
+  }), []);
+
+  const registeredOverlayKeys = Object.keys(loadingIndicators);
+  const overlay = (
+    <LoadingOverlay
+      isRelativeToContainer
+      isAnimated
+      isOpen={!!registeredOverlayKeys.length}
+      message={Object.keys(loadingIndicators).map(key => (
```

Comment:
This feels a little weird to me. I seems like an overlay can be registered and then unregistered in state via the show/hide overlay functions, but why not just override the current message instead of join the messages together?

Reply:
Good question.

```
   LOP
    |
    A
  /   \
B       C
```

If component B renders an overlay, and component C also renders an overlay at the same time, we wouldn't want to overwrite the message, right? Whoever renders last is going to erase the message of the first (even when the first is still loading).

But tbh I'm not super happy about joining the messages together either. I would rather remove the messages altogether if we can. @neilpfeiffer any insight into how we'd want to handle something like this?

Reply:
How did we decide to handle this?

Reply:
The messages were removed entirely due to future UX recommendations.